File: graphos-cpp/src/kernel/runtime.cpp
#include "graphos/kernel/runtime.hpp"
#include <openvino/openvino.hpp>
#include <chrono>
#include <mutex>
#include <unordered_map>

namespace graphos {

struct KernelRuntime::Impl {
    ov::Core core;
    std::string device;
    std::unordered_map<std::string, Program> programs;
    mutable std::recursive_mutex mu; // reentrant (execute calls get)

    // Metrics
    size_t exec_count = 0;
    double total_exec_us = 0.0;
    double last_exec_us = 0.0;
    size_t errors = 0;

    explicit Impl(const std::string& dev) {
        auto available = core.get_available_devices();
        bool found = false;
        for (const auto& d : available) {
            if (d == dev) { found = true; break; }
        }
        if (found) {
            device = dev;
        } else {
            device = "CPU";
        }
    }
};

KernelRuntime::KernelRuntime(const std::string& device)
    : impl_(std::make_unique<Impl>(device)) {}

KernelRuntime::~KernelRuntime() = default;
KernelRuntime::KernelRuntime(KernelRuntime&&) noexcept = default;
KernelRuntime& KernelRuntime::operator=(KernelRuntime&&) noexcept = default;

void KernelRuntime::load(const ProgramSpec& spec) {
    std::lock_guard lock(impl_->mu);
    if (impl_->programs.count(spec.name))
        throw KernelError("Program '" + spec.name + "' already loaded");

    auto model = impl_->core.read_model(spec.onnx_path);
    auto compiled = impl_->core.compile_model(model, impl_->device);

    // Program takes ownership via void* (pimpl boundary)
    impl_->programs.emplace(spec.name,
                            Program(spec, &compiled));
}

void KernelRuntime::unload(const std::string& name) {
    std::lock_guard lock(impl_->mu);
    auto it = impl_->programs.find(name);
    if (it == impl_->programs.end())
        throw KernelError("Program '" + name + "' not found");
    impl_->programs.erase(it);
}

bool KernelRuntime::has(const std::string& name) const {
    std::lock_guard lock(impl_->mu);
    return impl_->programs.count(name) > 0;
}

std::vector<std::string> KernelRuntime::programs() const {
    std::lock_guard lock(impl_->mu);
    std::vector<std::string> names;
    names.reserve(impl_->programs.size());
    for (const auto& [name, _] : impl_->programs)
        names.push_back(name);
    return names;
}

void KernelRuntime::execute(const std::string& name,
                            const float* input, float* output,
                            size_t batch_size) {
    Program* prog;
    {
        std::lock_guard lock(impl_->mu);
        auto it = impl_->programs.find(name);
        if (it == impl_->programs.end())
            throw KernelError("Program '" + name + "' not found");
        prog = &it->second;
    }

    auto t0 = std::chrono::steady_clock::now();
    try {
        prog->execute(input, output, batch_size);
    } catch (...) {
        std::lock_guard lock(impl_->mu);
        impl_->errors++;
        throw;
    }
    auto t1 = std::chrono::steady_clock::now();
    double us = std::chrono::duration<double, std::micro>(t1 - t0).count();

    std::lock_guard lock(impl_->mu);
    impl_->exec_count++;
    impl_->total_exec_us += us;
    impl_->last_exec_us = us;
}

const std::string& KernelRuntime::device() const noexcept {
    return impl_->device;
}

std::vector<std::string> KernelRuntime::available_devices() const {
    return impl_->core.get_available_devices();
}

HealthStatus KernelRuntime::health() const {
    std::lock_guard lock(impl_->mu);
    HealthStatus h;
    h.device = impl_->device;
    for (const auto& [name, _] : impl_->programs)
        h.programs.push_back(name);
    h.exec_count = impl_->exec_count;
    h.mean_latency_us = impl_->exec_count > 0
        ? impl_->total_exec_us / impl_->exec_count : 0.0;
    h.last_latency_us = impl_->last_exec_us;
    h.errors = impl_->errors;
    h.healthy = (impl_->errors == 0);
    return h;
}

} // namespace graphos

File: kernel/runtime.py
"""KernelRuntime — multi-program NPU/CPU execution environment."""

import threading
import time

import numpy as np

from kernel.program import ProgramSpec, Program


class KernelError(Exception):
    """Error raised by the kernel runtime."""


class KernelRuntime:
    """Manages device, loads and executes ONNX programs.

    Thread-safe. Falls back from NPU to CPU if NPU is unavailable.
    """

    def __init__(self, device: str = "NPU"):
        import openvino as ov

        self._ov = ov
        self._core = ov.Core()
        self._lock = threading.RLock()
        self._programs: dict[str, Program] = {}

        available = self._core.available_devices
        if device not in available:
            fallback = "CPU"
            print(f"WARNING: {device} not available (found: {available}). "
                  f"Falling back to {fallback}.")
            device = fallback
        self._device = device

        # Metrics
        self._exec_count = 0
        self._total_exec_us = 0.0
        self._last_exec_us = 0.0
        self._errors = 0

    @property
    def device(self) -> str:
        return self._device

    @property
    def programs(self) -> list[str]:
        """Names of loaded programs."""
        with self._lock:
            return list(self._programs.keys())

    def load(self, spec: ProgramSpec) -> Program:
        """Compile and register a program from its spec."""
        with self._lock:
            if spec.name in self._programs:
                raise KernelError(f"Program '{spec.name}' already loaded")
            model = self._core.read_model(spec.onnx_path)
            compiled = self._core.compile_model(model, self._device)
            program = Program(spec, compiled, self._ov)
            self._programs[spec.name] = program
            return program

    def unload(self, name: str) -> None:
        """Remove a program from the registry."""
        with self._lock:
            if name not in self._programs:
                raise KernelError(f"Program '{name}' not loaded")
            del self._programs[name]

    def get(self, name: str) -> Program:
        """Look up a loaded program by name."""
        with self._lock:
            if name not in self._programs:
                raise KernelError(f"Program '{name}' not loaded")
            return self._programs[name]

    def execute(self, program_name: str, input_tensor: np.ndarray) -> np.ndarray:
        """Execute a loaded program with timing."""
        program = self.get(program_name)
        t0 = time.perf_counter()
        try:
            result = program.execute(input_tensor)
        except Exception:
            self._errors += 1
            raise
        elapsed_us = (time.perf_counter() - t0) * 1e6
        with self._lock:
            self._exec_count += 1
            self._total_exec_us += elapsed_us
            self._last_exec_us = elapsed_us
        return result

    def device_info(self) -> dict:
        """Return info about the active device."""
        return {
            "device": self._device,
            "available_devices": self._core.available_devices,
        }

    def health(self) -> dict:
        """Return runtime health metrics."""
        with self._lock:
            mean_us = (self._total_exec_us / self._exec_count
                       if self._exec_count > 0 else 0.0)
            return {
                "device": self._device,
                "programs": list(self._programs.keys()),
                "exec_count": self._exec_count,
                "mean_latency_us": mean_us,
                "last_latency_us": self._last_exec_us,
                "errors": self._errors,
                "healthy": self._errors == 0,
            }
